README.md
# fluent-agent-lite

* http://github.com/tagomoris/fluent-agent-lite

## DESCRIPTION

'fluent-agent-lite' is a log transfer agent, for Fluentd's 'forward' input.

This agent reads specified files, and sends each lines to fluentd servers. One log line will be packed one fluentd message, that has one attribute ('message' or specified in configuration) with entire line (not terminated by newline).

### VERSION

1.0 https://github.com/tagomoris/fluent-agent-lite/releases/tag/v1.0

## PACKAGING TARBALL

To get tarball of latest tagged version:

    make -f Makefile.package

Got tarball on `tmp/archive/fluent-agent-lite.vX.Y.tar.gz`.

To do this, versions (tags) MUST be v1.0 not but v0.10 for sorting order...

## INSTALL

On RHEL/CentOS, you can use .spec file to build rpm with your customized default config file.

### RHEL/CentOS

To build your rpm package, do 5 steps below.

1. Download (newest version) tarball, and place it on SOURCES/ .
2. Download package/fluent-agent-lite.conf, and place it on SOURCES/ .
3. Fix SOURCES/package/fluent-agent-lite.conf as you want (ex: server name or servers list), and add servers list for your own.
4. Download SPECS/fluent-agent-lite.spec, and place it on SPECS/ .
5. run 'rpmbuild -ba SPECS/fluent-agent-lite.spec'

To install each RHEL/CentOS host, use yum server, or copy and rpm -i on each host.

NOTE: `yum install perl-devel` and `QA_RPATHS=$[0x001] rpmbuild -ba` may help you if `rpmbuild` fails on your build environment.

### Other Linux or Unix-like OS

On each host, do steps below.

1. Download and extract tarball, or clone repository, and move into extracted directory.
2. Do 'bin/install.sh'.

## Configuration

All of configurations are written in configuration shell-script file (/etc/fluent-agent-lite.conf). Configurable values are below:

### LOGS

Pairs of tag and file, such as:

    LOGS=$(cat <<"EOF"
    www     /var/log/nginx/www_access.log
    app     /var/log/apache2/app_access.log
    EOF

Or, you can use this syntax:

    LOGS=$(cat /etc/fluent-agent.logs)
    
    ## in fluent-agent.logs
    www  /var/log/nginx/www_access.log
    app  /var/log/apache2/app_access.log

### TAG_PREFIX

Prefix of each tags, specified in 'LOGS'. 'TAG_PREFIX="service"' with 'LOGS' above, you will get fluentd messages with tags 'service.www' and 'service.app'.

### FIELD_NAME

Log line attribute name in fluentd message (default: 'message').

### PRIMARY_SERVER, SECONDARY_SERVER

Fluentd server name and port (SERVERNAME:PORT), as primary server. 'fluent-agent-lite' try to connect to primary server at first, and if fails, then try to connect secondary server (if it specified).

Default port is 24224 (if omitted).

### PRIMARY_SERVERS_LIST, SECONDARY_SERVERS_LIST

File path to specify primary(secondary) servers' list. 'fluent-agent-lite' reads this file when executed, and choose one of servers randomly at each connection trial.

You cannot specify both of 'PRIMARY\_SERVER' and 'PRIMARY\_SERVERS\_LIST', and both of 'SECONDARY\_SERVER' and 'SECONDARY\_SERVERS\_LIST'

### READ_BUFFER_SIZE

Bytes size which 'fluent-agent-lite' try to read from 'tail' at once (Default: 1MB).

### PROCESS_NICE

Nice value for 'fluent-agent-lite'. If you want to execute fluent-agent-lite in server with high loadavg, 'PROCESS_NICE="-1"' can help you.

### TAIL_PATH

Path of tail command (Default: /usr/bin/tail).

### TAIL_INTERVAL

'sleep interval' of tail command in seconds (Default: 1.0). For high throughput log file, you can specify 'TAIL_INTERVAL="0.5"' or any other values (but over "0.1").

Caution: This cofiguration is for GNU tail only.

### PING_TAG, PING_DATA, PING_INTERVAL

Ping message tag/data and emit interval specification. Without PING\_TAG, fluent-agent-lite doesn't emit ping\_messages.

Actual 'data' field of ping message is 'PING_DATA PATH\_OF\_INPUT\_FILE'.

### DRAIN_LOG_TAG

Tag name of drain\_log (messages count per drain/send to server), which is emitted to configured fluentd sever as fluentd message. Default is none (not to send drain\_log).

### KEEPALIVE_TIME

Connection keepalive time in seconds. 0 means infinity (Default: 1800, minimum: 120)

### LOG_PATH

Log file path for 'fluent-agent-lite' (Default: /tmp/fluent-agent.log).

### LOG_VERBOSE

If you specify 'LOG_VERBOSE="yes"', 'fluent-agent-lite' writes logs with level info/debug (Default: warn/crit only).

## Run

With properly configured '/etc/fluent-agent-lite.conf', you can run and stop all transfer agent as below.

    # /etc/init.d/fluent-agent-lite start
    # /etc/init.d/fluent-agent-lite restart
    # /etc/init.d/fluent-agent-lite stop

Check running processes:

    # /etc/init.d/fluent-agent-lite status

And reset fluentd connections of each processes. (To re-connect primary server instead of secondary server immediately.)

    # /etc/init.d/fluent-agent-lite reload

To reflect change of config file, you must do 'restart'.

* * * * *

## License

Copyright 2012- TAGOMORI Satoshi (tagomoris)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Changes
revision history for Perl module Fluent::AgentLite

{{$NEXT}}
   - add force start option (w/o reading log file)
   - archive file maintenance

0.9 Thu Jun 27 15:43:37 JST 2013
   - maintenance update (update of package scripts)

0.8 Sat Apr 13 15:24:26 JST 2013
   - first version for CPAN dist



lib/App/FluentAgentLite.pm
package App::FluentAgentLite;

use strict;
use warnings;
use English;
use Carp;

use Time::Piece;
use Log::Minimal;

use IO::Socket::INET;
use Data::MessagePack;

our $VERSION = '0.01';

use constant CONNECTION_KEEPALIVE => 1800; # 30min
use constant CONNECTION_KEEPALIVE_MARGIN_MAX => 30; # max 30sec

use constant RECONNECT_WAIT_MIN => 0.5;  # 0.5sec
use constant RECONNECT_WAIT_MAX => 3600; # 60min
use constant RECONNECT_WAIT_INCR_RATE => 1.5;

use constant SEND_RETRY_WAIT => 0.5; # 0.5sec (fixed)

sub new {
    my $this = shift;
    my ($terminated, $tailfd, $tag, $primary_servers, $secondary_servers, $configuration) = @_;
    my $self = {
        terminated => $terminated,
        tailfd => $tailfd,
        tag => $tag,
        servers => {
            primary => $primary_servers,
            secondary => $secondary_servers,
        },
        buffer_size => $configuration->{buffer_size}
    };

    srand (time ^ $PID ^ unpack("%L*", `ps axww | gzip`));

    bless $self, $this;
}

sub execute {
    my $msgpack_packer = Data::MessagePack->new
    my $msgpack_sock; #TODO
    while() {
        # sysread -> pack -> send loop
    }
}

1;


lib/Fluent/AgentLite.pm
package Fluent::AgentLite;

use strict;
use warnings;
use English;
use Carp;

use POSIX qw(:errno_h);

use Time::HiRes;

use Time::Piece;
use Log::Minimal;

use IO::Socket::INET;
use Data::MessagePack;

our $VERSION = '1.0';

use constant READ_WAIT => 0.1; # 0.1sec

use constant SOCKET_TIMEOUT => 5; # 5sec

use constant CONNECTION_KEEPALIVE_INFINITY => 0;
use constant CONNECTION_KEEPALIVE_TIME => 1800; # 30min
use constant CONNECTION_KEEPALIVE_MIN => 120; # min 2min
use constant CONNECTION_KEEPALIVE_MARGIN_MAX => 30; # max 30sec

use constant RECONNECT_WAIT_MIN => 0.5;  # 0.5sec
use constant RECONNECT_WAIT_MAX => 3600; # 60min
use constant RECONNECT_WAIT_INCR_RATE => 1.5;

use constant SEND_RETRY_MAX => 4;

sub connection_keepalive_time {
    my ($keepalive_time) = @_;
    $keepalive_time + int(CONNECTION_KEEPALIVE_MARGIN_MAX * 2 * rand()) - CONNECTION_KEEPALIVE_MARGIN_MAX;
}

sub new {
    my $this = shift;
    my ($tag, $primary_servers, $secondary_servers, $configuration) = @_;
    my $self = {
        tag => $tag,
        servers => {
            primary => $primary_servers,
            secondary => $secondary_servers,
        },
        buffer_size => $configuration->{buffer_size},
        ping_message => $configuration->{ping_message},
        drain_log_tag => $configuration->{drain_log_tag},
        keepalive_time => $configuration->{keepalive_time},
        output_format => $configuration->{output_format},
    };

    if (defined $self->{output_format} and $self->{output_format} eq 'json') {
        *pack = *pack_json;
        *pack_drainlog = *pack_drainlog_json;
    }

    srand (time ^ $PID ^ unpack("%L*", `ps axww | gzip`));

    bless $self, $this;
}

sub execute {
    my $self = shift;
    my $args = shift;

    my $fieldname = $args->{fieldname};
    my $tailfd = $args->{tailfd};

    my $check_terminated = ($args->{checker} || {})->{term} || sub { 0 };
    my $check_reconnect = ($args->{checker} || {})->{reconnect} || sub { 0 };

    my $packer = Data::MessagePack->new();

    my $reconnect_wait = RECONNECT_WAIT_MIN;

    my $last_ping_message = time;
    if ($self->{ping_message}) {
        $last_ping_message = time - $self->{ping_message}->{interval} * 2;
    }
    my $keepalive_time = CONNECTION_KEEPALIVE_TIME;
    if (defined $self->{keepalive_time}) {
        $keepalive_time = $self->{keepalive_time};
        if ($keepalive_time < CONNECTION_KEEPALIVE_MIN and $keepalive_time != CONNECTION_KEEPALIVE_INFINITY) {
            warnf 'Keepalive time setting is too short. Set minimum value %s', CONNECTION_KEEPALIVE_MIN;
            $keepalive_time = CONNECTION_KEEPALIVE_MIN;
        }
    }

    my $pending_packed;
    my $continuous_line;
    my $disconnected_primary = 0;
    my $expiration_enable = $keepalive_time != CONNECTION_KEEPALIVE_INFINITY;

    while(not $check_terminated->()) {
        # at here, connection initialized (after retry wait if required)

        # connect to servers
        my $primary = $self->choose($self->{servers}->{primary});
        my $secondary;

        my $sock = $self->connect($primary) unless $disconnected_primary;
        if (not $sock and $self->{servers}->{secondary}) {
            $secondary = $self->choose($self->{servers}->{secondary});
            $sock = $self->connect($self->choose($self->{servers}->{secondary}));
        }
        $disconnected_primary = 0;
        unless ($sock) {
            # failed to connect both of primary / secondary
            warnf 'failed to connect servers, primary: %s, secondary: %s', $primary, ($secondary || 'none');
            warnf 'waiting %s seconds to reconnect', $reconnect_wait;

            Time::HiRes::sleep($reconnect_wait);
            $reconnect_wait *= RECONNECT_WAIT_INCR_RATE;
            $reconnect_wait = RECONNECT_WAIT_MAX if $reconnect_wait > RECONNECT_WAIT_MAX;
            next;
        }

        # succeed to connect. set keepalive disconnect time
        my $connecting = $secondary || $primary;

        my $expired = time + connection_keepalive_time($keepalive_time) if $expiration_enable;
        $reconnect_wait = RECONNECT_WAIT_MIN;

        while(not $check_reconnect->()) {
            # connection keepalive expired
            if ($expiration_enable and time > $expired) {
                infof "connection keepalive expired.";
                last;
            }

            # ping message (if enabled)
            my $ping_packed = undef;
            if ($self->{ping_message} and time >= $last_ping_message + $self->{ping_message}->{interval}) {
                $ping_packed = $self->pack_ping_message($packer, $self->{ping_message}->{tag}, $self->{ping_message}->{data});
                $last_ping_message = time;
            }

            # drain (sysread)
            my $lines = 0;
            if (not $pending_packed) {
                my $buffered_lines;
                ($buffered_lines, $continuous_line, $lines) = $self->drain($tailfd, $continuous_line);

                if ($buffered_lines) {
                    $pending_packed = $self->pack($packer, $fieldname, $buffered_lines);
                    if ($self->{drain_log_tag}) {
                        $pending_packed .= $self->pack_drainlog($packer, $self->{drain_log_tag}, $lines);
                    }
                }
                if ($ping_packed) {
                    $pending_packed ||= '';
                    $pending_packed .= $ping_packed;
                }
                unless ($pending_packed) {
                    Time::HiRes::sleep READ_WAIT;
                    next;
                }
            }
            # send
            my $written = $self->send($sock, $pending_packed);
            unless ($written) { # failed to write (socket error).
                $disconnected_primary = 1 unless $secondary;
                last;
            }

            $pending_packed = undef;
        }
        if ($check_reconnect->()) {
            infof "SIGHUP (or SIGTERM) received";
            $disconnected_primary = 0;
            $check_reconnect->(1); # clear SIGHUP signal
        }
        infof "disconnecting to current server";
        if ($sock) {
            $sock->close;
            $sock = undef;
        }
        infof "disconnected.";
    }
    if ($check_terminated->()) {
        warnf "SIGTERM received";
    }
    infof "process exit";
}

sub drain {
    # if broken child process (undefined return value of $fd->sysread())
    #   if content exists, return it.
    #   else die
    my ($self,$fd, $continuous_line) = @_;
    my $readlimit = $self->{buffer_size};
    my $readsize = 0;
    my $readlines = 0;
    my @buffered_lines;

    my $chunk;
    while ($readsize < $readlimit) {
        my $bytes = $fd->sysread($chunk, $readlimit);
        if (defined $bytes and $bytes == 0) { # EOF (child process exit)
            last if $readsize > 0;
            warnf "failed to read from child process, maybe killed.";
            confess "give up to read tailing fd, see logs";
        }
        if (not defined $bytes and $! == EAGAIN) { # I/O Error (no data in fd): "Resource temporarily unavailable"
            last;
        }
        if (not defined $bytes) { # Other I/O error... what?
            warnf "I/O error with tail fd: $!";
            last;
        }

        $readsize += $bytes;
        my $terminated_line = chomp $chunk;
        my @lines = split(m!\n!, $chunk);
        if ($continuous_line) {
            $lines[0] = $continuous_line . $lines[0];
            $continuous_line = undef;
        }
        if (not $terminated_line) {
            $continuous_line = pop @lines;
        }
        if (scalar(@lines) > 0) {
            push @buffered_lines, @lines;
            $readlines += scalar(@lines);
        }
    }
    if ($readlines < 1) {
        return undef, $continuous_line, 0;
    }

    return (\@buffered_lines, $continuous_line, $readlines);
}

# MessagePack 'PackedForward' object
# see lib/fluent/plugin/in_forward.rb in fluentd
sub pack {
    my ($self,$packer,$fieldname,$lines) = @_;
    my $t = time;
    my $event_stream = join('', map { $packer->pack([$t, {$fieldname => $_}]) } @$lines);
    return $packer->pack([$self->{tag}, $event_stream]);
}

sub pack_json {
    use JSON::XS qw/encode_json/;
    my ($self,$packer,$fieldname,$lines) = @_;
    my $t = time;
    return encode_json([$self->{tag}, [map { [$t, {$fieldname => $_}] } @$lines ]]);
}

# MessagePack 'Message' object
sub pack_ping_message {
    my ($self,$packer,$ping_tag,$ping_data) = @_;
    my $t = time;
    return $packer->pack([$ping_tag, $t, {'data' => $ping_data}]);
}

# MessagePack 'Message' object
sub pack_drainlog {
    my ($self,$packer,$drain_log_tag,$drain_lines) = @_;
    my $t = time;
    return $packer->pack([$drain_log_tag, $t, {'drain' => $drain_lines}]);
}

sub pack_drainlog_json {
    use JSON::XS qw/encode_json/;
    my ($self,$packer,$drain_log_tag,$drain_lines) = @_;
    my $t = time;
    return encode_json([$drain_log_tag, $t, {'drain' => $drain_lines}]);
}

# choose a server [host, port] randomly from arg arrayref
sub choose {
    my ($self,$servers) = @_;
    my $num = scalar(@$servers);
    return $servers->[int(rand() * $num)];
}

sub connect {
    my ($self,$server) = @_;
    my $sock = IO::Socket::INET->new(
        PeerAddr  => $server->[0],
        PeerPort  => $server->[1],
        Proto     => 'tcp',
        Timeout   => SOCKET_TIMEOUT,
        ReuseAddr => 1,
    );
    if ($sock) {
        infof 'connected to server: %s', $server;
    } else {
        warnf 'failed to connect to server %s : %s', $server, $!;
    }
    $sock;
}

sub send {
    my ($self,$sock,$data) = @_;
    my $length = length($data);
    my $written = 0;
    my $retry = 0;

    local $SIG{"PIPE"} = sub { die $! };

    eval {
        while ($written < $length) {
            my $wbytes = $sock->syswrite($data, $length, $written);
            if ($wbytes) {
                $written += $wbytes;
            }
            else {
                die "failed $retry times to send data: $!" if $retry > SEND_RETRY_MAX;
                $retry += 1;
            }
        }
    };
    if ($@) {
        my $error = $@;
        warnf "Cannot send data: $error";
        return undef;
    }
    $written;
}

sub close {
    my ($self,$sock) = @_;
    $sock->close if $sock;
}

1;
